doc/decisions/iterating_name_parts.md

```diff
@@ -0,0 +1,57 @@
+# Iterating Keyname Parts
+
+## Problem
+
+While iterating over the parts of a keyname is easy for an experienced developer (just jump from `\0` to `\0` until you hit the name size), there is no good API that newcomers can use.
+
+## Constraints
+
+No changes to the underlying APIs, while maintaining reasonable performance.
+
+## Assumptions
+
+## Considered Alternatives
+
+- Put the new function into `libelektra-core`.
+
+## Decision
+
+Add this new function to `libelektra-operations` (see [Operations Library](operations_library.md)):
+
+```c
+/**
+ * Iterate over the key name parts of @p key.
+ *
+ * @pre @p currentPart, must be `NULL` or point anywhere into the unescaped name of @p key
+ *
+ * This function defins the "next part" after @p currentPart as:
+ * - the first part of @p key (after namespace), if `currentPart == NULL`
+ * - `NULL`, if @p currentPart is the last part of @p key, i.e. `currentPart == keyBaseName (key)`
+ * - the next part within @p key after the location @p currentPart points to
+ *
+ * To iterate over the parts of a key you can use:
+ * @code{.c}
+ * Key * k;
+ * for (const char * part = keyGetNextPart (k, NULL); part != NULL; part = keyGetNextPart (k, part))
+ * {
+ *    // use part
+ * }
+ * @endcode
+ *
+ *
+ * @returns the "next" (see above) key name part after @p currentPart.
+ */
+const char * keyGetNextPart (Key * key, const char * currentPart);
+```
+
+## Rationale
+
+The function is not required for a minimal API, but it is useful for people who don't know everything about the internals of Elektra.
```

Comment:
I wouldn't see "uses private/internal API" as the boundary between `ease` and `operations`. I'd say `operations` contains more generally useful functions than `ease`. If we used "uses private/internal API" to determine where a function goes, we'd probably end up with one huge library and one tiny one. We'd also probably put functions that are closely related to each other in different libraries, just because one of them uses private APIs and the other doesn't.

However, `keyGetNextPart` should probably be in `libelektra-lowlevel-c` since it's API is very C-specific and other languages will use different solutions (mostly iterator based).

Reply:
> I'd say operations contains more generally useful functions than ease

I think this is too general and cannot be a guideline what the difference between different libraries is.

> we'd probably end up with one huge library and one tiny one.

I agree, we do not want that.

> keyGetNextPart in `libelektra-lowlevel-c`

Yes, I agree. We also need to watch out that `libelektra-lowlevel-c` doesn't grow huge as there are many helpers specific to C (and we probably do not want all of them in one huge library).

The next step would be to list all libraries and give guidelines what should be in which library.

Reply:
> The next step would be to list all libraries and give guidelines what should be in which library.

If we continue like this we will never get anywhere with these new libraries. We'll spend months discussing back and forth what the guidelines for each library are and not actually get anything done.

And maybe it's a typo, but if anything, I would say we need to list all _functions_ and then divide them into libraries. Listing the libraries and coming up with guidelines what each should contain seems very backwards to me.

---

It seems you want guidelines that clearly and with little doubt divide the set of functions Elektra provides into separate libraries, so that for new functions it's always clear where they belong. IMO it is almost impossible to come up with such guidelines, because there will always be a future function that is in between to libraries.

I think we shouldn't spend to much time coming up with these perfect guidelines. We will just have to adjust them, when we find a function that doesn't fit. I know I've suggested this before, but I think we should instead just go with a small set of clearly different libraries (`libelektra-lowlevel-c` is stuff that's only for C, `libelektra-operations` is stuff for everyone). Since we will still have separate libraries when a set of functions has a clear and separate use-case (e.g. `libelektra-opts`, `libelektra-merge`, `libelektra-notification`, etc.), I don't think the catch-all helper library `libelektra-operations` will grow that fast. By the time the library needs a split, we will probably have other reasons to make a breaking change release 2.0.

---

My main point is: IMO the best way to move forward is just "`libelektra-lowlevel-c` is stuff that's only for C, `libelektra-operations` is stuff for everyone" for now and re-examine the situation shortly before 1.0. Dividing up the libraries isn't too hard. Since the implementations wouldn't change, you just need to replace a few `#include`s and adjust some CMake configs.

Reply:
> > "is stuff for everyone"
> 
> This is too vague.

This wasn't intended as a definition. I just wanted to convey the general idea.

> Please think about a sentence for each library and write this down in a PR.

I added the `doc/dev/libraries.md` file in #4245. I incorporated all the current libraries, as well as the coming changes to the mount tooling in new-backend. The formatting still needs some work, but the general ideas should be good enough for review.